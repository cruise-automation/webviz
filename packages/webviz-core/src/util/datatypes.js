// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import type { MessageDefinitionsByTopic, ParsedMessageDefinitionsByTopic } from "webviz-core/src/players/types";
import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";
import { isComplex } from "webviz-core/src/util/binaryObjects/messageDefinitionUtils";
import {
  FUTURE_VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY,
  VISUALIZATION_MSGS$WEBVIZ_MARKER,
  VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY,
} from "webviz-core/src/util/globalConstants";

// Returns the subset of allDatatypes needed to fully define datatypes.
// If datatypes is ["visualization_msgs/Marker"], the returned set will include definitions of
// "std_msgs/Header", "geometry_msgs/Pose" etc, but not "visualization_msgs/MarkerArray".
export const getTransitiveSubsetForDatatypes = (
  allDatatypes: RosDatatypes,
  datatypes: $ReadOnlyArray<string>
): RosDatatypes => {
  // Breadth-first search. datatypesToExplore holds datatypes which will be included in the return
  // value, but whose children have not yet been visited.
  let datatypesToExplore = new Set(datatypes);
  const ret = {};
  while (datatypesToExplore.size > 0) {
    const nextDatatypesToExplore = new Set();
    for (const datatype of datatypesToExplore) {
      const definition = allDatatypes[datatype];
      if (definition == null) {
        throw new Error(`Definition missing for type ${datatype}`);
      }
      ret[datatype] = definition;
      for (const field of definition.fields) {
        if (isComplex(field.type) && ret[field.type] == null) {
          nextDatatypesToExplore.add(field.type);
        }
      }
    }
    datatypesToExplore = nextDatatypesToExplore;
  }
  return ret;
};

const markerFields = [
  { type: "uint8", name: "ARROW", isConstant: true, value: 0 },
  { type: "uint8", name: "CUBE", isConstant: true, value: 1 },
  { type: "uint8", name: "SPHERE", isConstant: true, value: 2 },
  { type: "uint8", name: "CYLINDER", isConstant: true, value: 3 },
  { type: "uint8", name: "LINE_STRIP", isConstant: true, value: 4 },
  { type: "uint8", name: "LINE_LIST", isConstant: true, value: 5 },
  { type: "uint8", name: "CUBE_LIST", isConstant: true, value: 6 },
  { type: "uint8", name: "SPHERE_LIST", isConstant: true, value: 7 },
  { type: "uint8", name: "POINTS", isConstant: true, value: 8 },
  { type: "uint8", name: "TEXT_VIEW_FACING", isConstant: true, value: 9 },
  { type: "uint8", name: "MESH_RESOURCE", isConstant: true, value: 10 },
  { type: "uint8", name: "TRIANGLE_LIST", isConstant: true, value: 11 },
  { type: "uint8", name: "ADD", isConstant: true, value: 0 },
  { type: "uint8", name: "MODIFY", isConstant: true, value: 0 },
  { type: "uint8", name: "DELETE", isConstant: true, value: 2 },
  { type: "std_msgs/Header", name: "header", isArray: false, isComplex: true },
  { type: "string", name: "ns", isArray: false, isComplex: false },
  { type: "int32", name: "id", isArray: false, isComplex: false },
  { type: "int32", name: "type", isArray: false, isComplex: false },
  { type: "int32", name: "action", isArray: false, isComplex: false },
  { type: "geometry_msgs/Pose", name: "pose", isArray: false, isComplex: true },
  { type: "geometry_msgs/Vector3", name: "scale", isArray: false, isComplex: true },
  { type: "std_msgs/ColorRGBA", name: "color", isArray: false, isComplex: true },
  { type: "duration", name: "lifetime", isArray: false, isComplex: false },
  { type: "bool", name: "frame_locked", isArray: false, isComplex: false },
  { type: "geometry_msgs/Point", name: "points", isArray: true, arrayLength: undefined, isComplex: true },
  { type: "std_msgs/ColorRGBA", name: "colors", isArray: true, arrayLength: undefined, isComplex: true },
  { type: "string", name: "text", isArray: false, isComplex: false },
  { type: "string", name: "mesh_resource", isArray: false, isComplex: false },
  { type: "bool", name: "mesh_use_embedded_materials", isArray: false, isComplex: false },
];

export const basicDatatypes: RosDatatypes = {
  [FUTURE_VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY]: {
    fields: [
      { type: "std_msgs/Header", name: "header", isArray: false, isComplex: true },
      {
        isArray: true,
        isComplex: true,
        name: "allMarkers",
        type: VISUALIZATION_MSGS$WEBVIZ_MARKER,
        arrayLength: undefined,
      },
    ],
  },
  [VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY]: {
    fields: [
      {
        isArray: true,
        isComplex: true,
        arrayLength: undefined,
        name: "markers",
        type: VISUALIZATION_MSGS$WEBVIZ_MARKER,
      },
      {
        isArray: false,
        isComplex: true,
        name: "header",
        type: "std_msgs/Header",
      },
    ],
  },
  "visualization_msgs/MarkerArray": {
    fields: [
      {
        isArray: true,
        isComplex: true,
        arrayLength: undefined,
        name: "markers",
        type: "visualization_msgs/Marker",
      },
    ],
  },
  "visualization_msgs/Marker": { fields: markerFields },
  // This is a special marker type that has a string instead of an int ID field and an additional JSON "metadata" field.
  // For use internally to webviz, when we need to add extra data to markers.
  [VISUALIZATION_MSGS$WEBVIZ_MARKER]: {
    fields: markerFields
      .filter(({ name }) => name !== "id")
      .concat([
        { type: "string", name: "id", isArray: false, isComplex: false },
        { type: "json", name: "metadata", isArray: false, isComplex: false },
      ]),
  },
  "std_msgs/ColorRGBA": {
    fields: [
      { type: "float32", name: "r", isArray: false, isComplex: false },
      { type: "float32", name: "g", isArray: false, isComplex: false },
      { type: "float32", name: "b", isArray: false, isComplex: false },
      { type: "float32", name: "a", isArray: false, isComplex: false },
    ],
  },
  "std_msgs/Header": {
    fields: [
      { type: "uint32", name: "seq", isArray: false, isComplex: false },
      { type: "time", name: "stamp", isArray: false, isComplex: false },
      { type: "string", name: "frame_id", isArray: false, isComplex: false },
    ],
  },
  "geometry_msgs/Pose": {
    fields: [
      { type: "geometry_msgs/Point", name: "position", isArray: false, isComplex: true },
      { type: "geometry_msgs/Quaternion", name: "orientation", isArray: false, isComplex: true },
    ],
  },
  "geometry_msgs/PoseStamped": {
    fields: [
      {
        isArray: false,
        isComplex: true,
        name: "header",
        type: "std_msgs/Header",
      },
      {
        isArray: false,
        isComplex: true,
        name: "pose",
        type: "geometry_msgs/Pose",
      },
    ],
  },

  "geometry_msgs/Vector3": {
    fields: [
      { type: "float64", name: "x", isArray: false, isComplex: false },
      { type: "float64", name: "y", isArray: false, isComplex: false },
      { type: "float64", name: "z", isArray: false, isComplex: false },
    ],
  },
  "geometry_msgs/Point": {
    fields: [
      { type: "float64", name: "x", isArray: false, isComplex: false },
      { type: "float64", name: "y", isArray: false, isComplex: false },
      { type: "float64", name: "z", isArray: false, isComplex: false },
    ],
  },
  "geometry_msgs/Quaternion": {
    fields: [
      { type: "float64", name: "x", isArray: false, isComplex: false },
      { type: "float64", name: "y", isArray: false, isComplex: false },
      { type: "float64", name: "z", isArray: false, isComplex: false },
      { type: "float64", name: "w", isArray: false, isComplex: false },
    ],
  },
};

let datatypeSetsGenerated = 0;
function getGetDatatypeName() {
  const prefix = `f_${datatypeSetsGenerated++}`;
  let count = 0;
  const names = {};
  return (key) => {
    if (names[key] == null) {
      names[key] = `${prefix}_${count++}`;
    }
    return names[key];
  };
}

export const resetDatatypePrefixForTest = () => {
  datatypeSetsGenerated = 0;
};

const renameDatatypes = (datatypes: RosDatatypes, typeName: string, getNewName) => {
  // Generate name/id mappings.
  const nameMapping = {};
  Object.keys(datatypes).forEach((datatype) => {
    const childDatatypes = Object.keys(getTransitiveSubsetForDatatypes(datatypes, [datatype])).sort();
    const typeKey = childDatatypes.map((childType) => JSON.stringify(datatypes[childType])).join("\n");
    nameMapping[datatype] = getNewName(typeKey);
  });

  // Generate mapped datatype definitions.
  const idMappedDatatypes = {};
  Object.keys(datatypes).forEach((datatype) => {
    idMappedDatatypes[nameMapping[datatype]] = {
      fields: datatypes[datatype].fields.map((field) => ({
        ...field,
        type: nameMapping[field.type] ?? field.type,
      })),
    };
  });

  return { idMappedDatatypes, topicDatatypeId: nameMapping[typeName] };
};

// Replace all type names with names that depend only on structural properties.
// If two messages have the same fields they should get the same name.
// If they have different fields (or even if their children differ in any way) they should get
// different names.
export const getContentBasedDatatypes = (
  messageDefinitionsByTopic: MessageDefinitionsByTopic,
  parsedMessageDefinitionsByTopic: ParsedMessageDefinitionsByTopic,
  datatypesByTopic: { [topic: string]: string }
): { fakeDatatypesByTopic: { [topic: string]: string }, fakeDatatypes: RosDatatypes } => {
  // Parsing message definitions is expensive, but topics often share message definitions.
  // We only compute new datatypes for unique string message definitions, and copy the result to
  // topics with those definitions.
  const topicsByStringDefinition = {};
  const fakeDatatypesByStringDefinition = {};
  const allFakeDatatypes = {};
  const getDatatypeName = getGetDatatypeName();

  // The string message definitions by topic could be incomplete, so use the parsed ones to turn into datatypes.
  Object.keys(parsedMessageDefinitionsByTopic).forEach((topic) => {
    const definition = messageDefinitionsByTopic[topic];
    const parsedDefinition = parsedMessageDefinitionsByTopic[topic];
    // This "key" is just used to group topics with identical sets of datatypes.
    const stringDefinitionKey = definition != null ? definition : JSON.stringify(parsedDefinition);
    if (topicsByStringDefinition[stringDefinitionKey]) {
      // Already generated datatypes for these message types.
      topicsByStringDefinition[stringDefinitionKey].push(topic);
    } else {
      // First time seeing this definition. Generate new datatypes for it.
      topicsByStringDefinition[stringDefinitionKey] = [topic];

      // Convert parsedDefinition:RosMsgField[] to datatypes:RosDatatypes
      const datatypes = {};
      parsedDefinition.forEach((datatype) => {
        const typeName = datatype.name ?? datatypesByTopic[topic];
        datatypes[typeName] = { fields: datatype.definitions };
      });

      // Rename datatypes for this set of message definitions (reusing types we've seen before.)
      const { idMappedDatatypes, topicDatatypeId } = renameDatatypes(
        datatypes,
        datatypesByTopic[topic],
        getDatatypeName
      );
      Object.assign(allFakeDatatypes, idMappedDatatypes);
      fakeDatatypesByStringDefinition[stringDefinitionKey] = topicDatatypeId;
    }
  });

  // Transform "per stringified set of message definitions" output to "per topic".
  const fakeDatatypesByTopic = {};
  Object.keys(topicsByStringDefinition).forEach((stringDefinition) => {
    topicsByStringDefinition[stringDefinition].forEach((topic) => {
      fakeDatatypesByTopic[topic] = fakeDatatypesByStringDefinition[stringDefinition];
    });
  });

  return { fakeDatatypesByTopic, fakeDatatypes: allFakeDatatypes };
};
