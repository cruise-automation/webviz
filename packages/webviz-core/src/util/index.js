// @flow
//
//  Copyright (c) 2019-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.
import { debounce } from "lodash";

export function arrayToPoint(v: ?[number, number, number]) {
  if (!v) {
    return null;
  }
  return { x: v[0], y: v[1], z: v[2] };
}

// returns the linear interpolation between a and b based on unit-range variable t
export function lerp(t: number, a: number, b: number): number {
  // Clamp t to (0, 1)
  t = Math.min(Math.max(t, 0.0), 1.0);
  if (a === b) {
    return a;
  }
  return a + t * (b - a);
}

// the following regex captures characters allowed in the value of a kv-pair in the query component
// of a URI, minus "&" and "+" because they are handled specially by browsers.
//   https://tools.ietf.org/html/rfc3986
//   query = *( pchar / "/" / "?" )
//   pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
const QUERY_REGEXP = /([a-zA-Z0-9\-._~!$'()*,;=:@/?])|./gu;

// percent-encode a parameter for the query portion of a URL.
// leaves certain characters un-escaped that the URLSearchParams class does not,
// to improve readability of the URL.
export function encodeURLQueryParamValue(value: string): string {
  return value.replace(QUERY_REGEXP, (char, allowedChar) => {
    return allowedChar || encodeURIComponent(char);
  });
}

// extra boundary added for jest testing, since jsdom's Blob doesn't support .text()
export function downloadTextFile(text: string, fileName: string) {
  return downloadFiles([{ blob: new Blob([text]), fileName }]);
}

export function downloadFiles(files: { blob: Blob, fileName: string }[]) {
  const { body } = document;
  if (!body) {
    return;
  }

  const link = document.createElement("a");
  link.style.display = "none";
  body.appendChild(link);

  const urls = files.map((file) => {
    const url = window.URL.createObjectURL(file.blob);
    link.setAttribute("download", file.fileName);
    link.setAttribute("href", url);
    link.click();
    return url;
  });

  // remove the link after triggering download
  window.requestAnimationFrame(() => {
    body.removeChild(link);
    urls.forEach((url) => {
      URL.revokeObjectURL(url);
    });
  });
}

// Equivalent to `number % modulus`, but always returns a positive number (given that modulus is
// a positive number). This is the same as the `%` in e.g. Python.
// See https://stackoverflow.com/a/4467559 and https://en.wikipedia.org/wiki/Modulo_operation
export function positiveModulo(number: number, modulus: number): number {
  return ((number % modulus) + modulus) % modulus;
}

// Object.values returns mixed[], which is difficult to get Flow to accept.
export function objectValues<K, V>(o: $ReadOnly<{ [keys: K]: V }>): V[] {
  return (Object.values(o): any);
}

// Used when we want to limit some log rate, but the log contents depend on the combined contents
// of each event (like a sum).
export const debounceReduce = <A, T>({
  action, // Debounced function
  wait,
  reducer, // Combining/mapping function for action's argument
  initialValue,
}: {
  action: (A) => void,
  wait: number,
  reducer: (A, T) => A,
  initialValue: A,
}): ((T) => void) => {
  let current = initialValue;
  const debounced = debounce(
    () => {
      action(current);
      current = initialValue;
    },
    wait,
    { leading: true, trailing: true }
  );
  return (next: T) => {
    current = reducer(current, next);
    debounced();
  };
};
