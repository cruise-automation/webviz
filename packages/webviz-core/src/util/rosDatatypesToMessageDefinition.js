// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { type RosMsgDefinition } from "rosbag";

import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";

// For one datatype in the datatypes, find the RosMsgDefinition that we can use to either write or parse it.
// `datatypes` should contain the root datatype and all complex sub-datatypes.
export default function rosDatatypesToMessageDefinition(
  datatypes: RosDatatypes,
  rootDatatypeName: string
): RosMsgDefinition[] {
  const result = [];
  const seenDatatypeNames = new Set([rootDatatypeName]);
  // It doesn't matter if we use a stack or queue here, but we use a stack.
  const datatypeNameStack = [rootDatatypeName];

  while (datatypeNameStack.length) {
    const currentDatatypeName = datatypeNameStack.pop();
    const currentDatatype = datatypes[currentDatatypeName];
    if (!currentDatatype) {
      throw new Error(
        `While searching datatypes for "${rootDatatypeName}", could not find datatype "${currentDatatypeName}"`
      );
    }
    // The root datatype has no name field.
    const msgDefinition: RosMsgDefinition =
      currentDatatypeName === rootDatatypeName
        ? { definitions: currentDatatype.fields }
        : { name: currentDatatypeName, definitions: currentDatatype.fields };
    result.push(msgDefinition);
    for (const field of currentDatatype.fields) {
      // Only search subfields if we haven't already seen it and it is "complex", IE it has its own fields and should
      // be contained in `datatypes`.
      if (field.isComplex && !seenDatatypeNames.has(field.type)) {
        datatypeNameStack.push(field.type);
        seenDatatypeNames.add(field.type);
      }
    }
  }

  return result;
}
