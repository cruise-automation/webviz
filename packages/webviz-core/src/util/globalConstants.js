// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.
import { colors } from "webviz-core/src/util/sharedStyleConstants";

// URL params
// DANGER: if you change this you break existing urls
export const REMOTE_BAG_URL_QUERY_KEY = "remote-bag-url";
export const REMOTE_BAG_URL_2_QUERY_KEY = "remote-bag-url-2";
export const ROSBRIDGE_WEBSOCKET_URL_QUERY_KEY = "rosbridge-websocket-url";
export const MEASURE_DATA_PROVIDERS_QUERY_KEY = "_measureDataProviders";
export const DEMO_QUERY_KEY = "demo";
export const DISABLE_WORKERS_QUERY_KEY = "no-workers";
export const SEEK_TO_QUERY_KEY = "seek-to"; // Used on load and set when paused
export const SEEK_TO_RELATIVE_MS_QUERY_KEY = "seek-by"; // Only used on load. Can be negative.
export const SEEK_TO_FRACTION_QUERY_KEY = "seek-fraction"; // Only used on load
export const LAYOUT_QUERY_KEY = "layout";
export const LAYOUT_URL_QUERY_KEY = "layout-url";
export const PATCH_QUERY_KEY = "patch";
export const OLD_GLOBAL_VARIABLES_QUERY_KEY = "global-data";
export const GLOBAL_VARIABLES_QUERY_KEY = "global-variables";
export const TITLE_QUERY_KEY = "title";
export const TEST_EVERYTHING_LAYOUT_QUERY_VAL = "_integration-test-everything-layout";
export const FRAMELESS = "frameless";

export const DEFAULT_WEBVIZ_NODE_PREFIX = "/webviz_node/";

export const TRANSFORM_TOPIC = "/tf";
export const TRANSFORM_STATIC_TOPIC = "/tf_static";
export const DIAGNOSTIC_TOPIC = "/diagnostics";
export const ROSOUT_TOPIC = "/rosout";
export const SOCKET_KEY = "dataSource.websocket";
export const SECOND_SOURCE_PREFIX = "/webviz_source_2";

export const GEOMETRY_MSGS$POLYGON_STAMPED = "geometry_msgs/PolygonStamped";
export const NAV_MSGS$OCCUPANCY_GRID = "nav_msgs/OccupancyGrid";
export const NAV_MSGS$PATH = "nav_msgs/Path";
export const SENSOR_MSGS$POINT_CLOUD_2 = "sensor_msgs/PointCloud2";
export const GEOMETRY_MSGS$POSE_STAMPED = "geometry_msgs/PoseStamped";
export const SENSOR_MSGS$LASER_SCAN = "sensor_msgs/LaserScan";
export const VISUALIZATION_MSGS$WEBVIZ_MARKER = "visualization_msgs/WebvizMarker";
export const VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY = "visualization_msgs/WebvizMarkerArray";
export const FUTURE_VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY = "future_visualization_msgs/WebvizMarkerArray";
export const TF2_MSGS$TF_MESSAGE = "tf2_msgs/TFMessage";
export const VISUALIZATION_MSGS$MARKER = "visualization_msgs/Marker";
export const VISUALIZATION_MSGS$MARKER_ARRAY = "visualization_msgs/MarkerArray";

export const WEBVIZ_ICON_MSGS$WEBVIZ_2D_ICON_ARRAY = "webviz_icon_msgs/WebViz2dIconArray";
export const WEBVIZ_ICON_MSGS$WEBVIZ_3D_ICON_ARRAY = "webviz_icon_msgs/WebViz3dIconArray";

export const MARKER_ARRAY_DATATYPES = [
  VISUALIZATION_MSGS$MARKER_ARRAY,
  FUTURE_VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY,
  VISUALIZATION_MSGS$WEBVIZ_MARKER_ARRAY,
];

export const USER_ERROR_PREFIX = "[WEBVIZ USER ERROR]";

// In testing there seems to be a small (~4%) performance penalty to freezing messages, so for now
// we only do it in development/testing. StackOverflow says that there should be no difference
// though: https://stackoverflow.com/questions/8435080/any-performance-benefit-to-locking-down-javascript-objects
// So maybe we want to always do the freezing at some point? Probably requires some more testing to
// be sure.
export const FREEZE_MESSAGES = process.env.NODE_ENV !== "production";

export const COLORS = {
  RED: { r: 1.0, g: 0.2, b: 0.2, a: 1.0 },
  BLUE: { r: 0.4, g: 0.4, b: 1.0, a: 1.0 },
  YELLOW: { r: 0.9, g: 1.0, b: 0.1, a: 1.0 },
  ORANGE: { r: 1.0, g: 0.6, b: 0.2, a: 1.0 },
  GREEN: { r: 0.1, g: 0.9, b: 0.3, a: 1.0 },
  GRAY: { r: 0.4, g: 0.4, b: 0.4, a: 1.0 },
  PURPLE: { r: 1.0, g: 0.2, b: 1.0, a: 1.0 },
  WHITE: { r: 1.0, g: 1.0, b: 1.0, a: 1.0 },
  PINK: { r: 1.0, g: 0.4, b: 0.6, a: 1.0 },
  LIGHT_RED: { r: 0.9, g: 0.1, b: 0.1, a: 1.0 },
  LIGHT_GREEN: { r: 0.4, g: 0.9, b: 0.4, a: 1.0 },
  LIGHT_BLUE: { r: 0.4, g: 0.4, b: 1, a: 1.0 },
  CLEAR: { r: 0, g: 0, b: 0, a: 0 },
};

// http://docs.ros.org/melodic/api/visualization_msgs/html/msg/Marker.html
export const MARKER_MSG_TYPES = {
  ARROW: 0,
  CUBE: 1,
  SPHERE: 2,
  CYLINDER: 3,
  LINE_STRIP: 4,
  LINE_LIST: 5,
  CUBE_LIST: 6,
  SPHERE_LIST: 7,
  POINTS: 8,
  TEXT_VIEW_FACING: 9,
  MESH_RESOURCE: 10,
  TRIANGLE_LIST: 11,
  FILLED_POLYGON: 107,
  INSTANCED_LINE_LIST: 108,
  OVERLAY_ICON: 109,
};

export const POSE_MARKER_SCALE = { x: 2, y: 2, z: 0.1 };

// Planning
export const METERS_PER_SECOND_TO_MILES_PER_HOUR = 2.23694;
export const METERS_PER_SECOND_TO_KILOMETERS_PER_HOUR = 3.6;
export const MILES_PER_HOUR_TO_METERS_PER_SECOND = 1 / METERS_PER_SECOND_TO_MILES_PER_HOUR;

export const jsonTreeTheme = {
  base00: "transparent", // bg
  base07: colors.BLUEL1, // text
  base0B: colors.YELLOW1, // string & date, item string
  base09: colors.REDL1, // # & boolean
  base08: colors.RED, // null, undefined, function, & symbol
  base0D: colors.BLUEL1, // label & arrow
  base03: colors.DARK9, // item string expanded
};

export const TAB_PANEL_TYPE = "Tab";

export const LINED_CONVEX_HULL_RENDERING_SETTING = "LinedConvexHull";

export const PANEL_LAYOUT_ROOT_ID = "PanelLayout-root";

// Feature announcements
export const FEATURE_ANNOUNCEMENTS_LOCAL_STORAGE_KEY = "webvizFeatureAnnouncements";
export const RECORDING_SERVICE_ANNOUNCEMENT_KEY = "recordingService";
