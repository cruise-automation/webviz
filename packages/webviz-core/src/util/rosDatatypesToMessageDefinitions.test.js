// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { uniqBy } from "lodash";

import { basicDatatypes } from "./datatypes";
import rosDatatypesToMessageDefinition from "./rosDatatypesToMessageDefinition";
import { WEBVIZ_MARKER_ARRAY_DATATYPE, WEBVIZ_MARKER_DATATYPE } from "webviz-core/src/util/globalConstants";

describe("rosDatatypesToMessageDefinition", () => {
  it(`Includes all of the definitions for "visualization_msgs/WebvizMarkerArray"`, () => {
    expect(rosDatatypesToMessageDefinition(basicDatatypes, WEBVIZ_MARKER_ARRAY_DATATYPE)).toMatchSnapshot();
  });

  it("produces a correct message definition", () => {
    const definitions = rosDatatypesToMessageDefinition(basicDatatypes, WEBVIZ_MARKER_ARRAY_DATATYPE);
    // Should have 1 definition without a name, the root datatype.
    expect(definitions.filter(({ name }) => !name).length).toEqual(1);
    // Should not duplicate definitions.
    expect(uniqBy(definitions, "name").length).toEqual(definitions.length);
  });

  it("Errors if it can't find the definition", () => {
    const datatypes = {
      [WEBVIZ_MARKER_ARRAY_DATATYPE]: {
        fields: [
          {
            isArray: true,
            isComplex: true,
            arrayLength: undefined,
            name: "markers",
            type: WEBVIZ_MARKER_DATATYPE,
          },
          {
            isArray: false,
            isComplex: true,
            name: "header",
            type: "std_msgs/Header",
          },
        ],
      },
    };
    expect(() => rosDatatypesToMessageDefinition(datatypes, WEBVIZ_MARKER_ARRAY_DATATYPE)).toThrow(
      `While searching datatypes for "visualization_msgs/WebvizMarkerArray", could not find datatype "std_msgs/Header"`
    );
  });
});
