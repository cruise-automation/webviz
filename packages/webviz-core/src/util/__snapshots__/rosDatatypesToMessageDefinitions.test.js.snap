// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rosDatatypesToMessageDefinition Includes all of the definitions for "visualization_msgs/WebvizMarkerArray" 1`] = `
Array [
  Object {
    "definitions": Array [
      Object {
        "arrayLength": undefined,
        "isArray": true,
        "isComplex": true,
        "name": "markers",
        "type": "visualization_msgs/WebvizMarker",
      },
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "header",
        "type": "std_msgs/Header",
      },
    ],
  },
  Object {
    "definitions": Array [
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "seq",
        "type": "uint32",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "stamp",
        "type": "time",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "frame_id",
        "type": "string",
      },
    ],
    "name": "std_msgs/Header",
  },
  Object {
    "definitions": Array [
      Object {
        "isConstant": true,
        "name": "ARROW",
        "type": "uint8",
        "value": 0,
      },
      Object {
        "isConstant": true,
        "name": "CUBE",
        "type": "uint8",
        "value": 1,
      },
      Object {
        "isConstant": true,
        "name": "SPHERE",
        "type": "uint8",
        "value": 2,
      },
      Object {
        "isConstant": true,
        "name": "CYLINDER",
        "type": "uint8",
        "value": 3,
      },
      Object {
        "isConstant": true,
        "name": "LINE_STRIP",
        "type": "uint8",
        "value": 4,
      },
      Object {
        "isConstant": true,
        "name": "LINE_LIST",
        "type": "uint8",
        "value": 5,
      },
      Object {
        "isConstant": true,
        "name": "CUBE_LIST",
        "type": "uint8",
        "value": 6,
      },
      Object {
        "isConstant": true,
        "name": "SPHERE_LIST",
        "type": "uint8",
        "value": 7,
      },
      Object {
        "isConstant": true,
        "name": "POINTS",
        "type": "uint8",
        "value": 8,
      },
      Object {
        "isConstant": true,
        "name": "TEXT_VIEW_FACING",
        "type": "uint8",
        "value": 9,
      },
      Object {
        "isConstant": true,
        "name": "MESH_RESOURCE",
        "type": "uint8",
        "value": 10,
      },
      Object {
        "isConstant": true,
        "name": "TRIANGLE_LIST",
        "type": "uint8",
        "value": 11,
      },
      Object {
        "isConstant": true,
        "name": "ADD",
        "type": "uint8",
        "value": 0,
      },
      Object {
        "isConstant": true,
        "name": "MODIFY",
        "type": "uint8",
        "value": 0,
      },
      Object {
        "isConstant": true,
        "name": "DELETE",
        "type": "uint8",
        "value": 2,
      },
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "header",
        "type": "std_msgs/Header",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "ns",
        "type": "string",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "type",
        "type": "int32",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "action",
        "type": "int32",
      },
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "pose",
        "type": "geometry_msgs/Pose",
      },
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "scale",
        "type": "geometry_msgs/Vector3",
      },
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "color",
        "type": "std_msgs/ColorRGBA",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "lifetime",
        "type": "duration",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "frame_locked",
        "type": "bool",
      },
      Object {
        "arrayLength": undefined,
        "isArray": true,
        "isComplex": true,
        "name": "points",
        "type": "geometry_msgs/Point",
      },
      Object {
        "arrayLength": undefined,
        "isArray": true,
        "isComplex": true,
        "name": "colors",
        "type": "std_msgs/ColorRGBA",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "text",
        "type": "string",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "mesh_resource",
        "type": "string",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "mesh_use_embedded_materials",
        "type": "bool",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "id",
        "type": "string",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "metadata",
        "type": "json",
      },
    ],
    "name": "visualization_msgs/WebvizMarker",
  },
  Object {
    "definitions": Array [
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "x",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "y",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "z",
        "type": "float64",
      },
    ],
    "name": "geometry_msgs/Point",
  },
  Object {
    "definitions": Array [
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "r",
        "type": "float32",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "g",
        "type": "float32",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "b",
        "type": "float32",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "a",
        "type": "float32",
      },
    ],
    "name": "std_msgs/ColorRGBA",
  },
  Object {
    "definitions": Array [
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "x",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "y",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "z",
        "type": "float64",
      },
    ],
    "name": "geometry_msgs/Vector3",
  },
  Object {
    "definitions": Array [
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "position",
        "type": "geometry_msgs/Point",
      },
      Object {
        "isArray": false,
        "isComplex": true,
        "name": "orientation",
        "type": "geometry_msgs/Quaternion",
      },
    ],
    "name": "geometry_msgs/Pose",
  },
  Object {
    "definitions": Array [
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "x",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "y",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "z",
        "type": "float64",
      },
      Object {
        "isArray": false,
        "isComplex": false,
        "name": "w",
        "type": "float64",
      },
    ],
    "name": "geometry_msgs/Quaternion",
  },
]
`;
